neuro_evolution_ctrnn/brains/ffnn.py

```diff
@@ -94,64 +72,56 @@ def __init__(self, input_space: Space, output_space: Space, individual: np.ndarr
-                    np.array(individual[current_index:current_index + hidden_layer]))
-                current_index += hidden_layer
-
-        # TODO this has to be redone if it shall be used
-        # # Indirect encoding
-        # if config.indirect_encoding:
-        #
-        #     config_cppn = LayeredNNCfg(number_neurons_layer1=config.cppn_hidden_size1,
-        #                                number_neurons_layer2=config.cppn_hidden_size2, indirect_encoding=False,
-        #                                use_biases=False, cppn_hidden_size1=0, cppn_hidden_size2=0, type="LNN")
-        #
-        #     cppn_weights = LayeredNN(4, 1, individual, config_cppn)
-        #
-        #     self.W1 = np.zeros((self.hidden_size1, self.input_size), dtype=np.single)
-        #     self.W2 = np.zeros((self.hidden_size2, self.hidden_size1), dtype=np.single)
-        #     self.W3 = np.zeros((self.output_size, self.hidden_size2), dtype=np.single)
-        #
-        #     for i, j in np.ndindex(self.W1.shape):
-        #         self.W1[i, j] = cppn_weights.step(
-        #             np.array([i / (self.hidden_size1 - 1), 0.33, j / (self.input_size - 1), 0]))
-        #
-        #     for i, j in np.ndindex(self.W2.shape):
-        #         self.W2[i, j] = cppn_weights.step(
-        #             np.array([i / (self.hidden_size2 - 1), 0.66, j / (self.hidden_size1 - 1), 0.33]))
-        #
-        #     for i, j in np.ndindex(self.W3.shape):
-        #         self.W3[i, j] = cppn_weights.step(
-        #             np.array([i / (self.output_size - 1), 1.0, j / (self.hidden_size2 - 1), 0.66]))
-        #
+            current_layer = nn.Linear(previous_layer_size, hidden_layer_size, bias=self.use_bias)
+            current_weight, index = self.read_matrix_from_genome(individual, index, hidden_layer_size, previous_layer_size)
+            current_layer.weight.data = torch.from_numpy(current_weight)
+
+            if self.use_bias:
+                current_bias, index = self.read_matrix_from_genome(individual, index, hidden_layer_size, 1)
+                current_layer.bias.data = torch.from_numpy(current_bias.flatten())
+
+            self.layers.append(current_layer)
+
+            previous_layer_size = hidden_layer_size
+
+        # Output Layer
+        self.output_layer: nn.Linear = nn.Linear(previous_layer_size, self.output_size, bias=config.use_bias)
+        weights_output_layer, index = self.read_matrix_from_genome(individual, index, self.output_size, previous_layer_size)
+        self.output_layer.weight.data = torch.from_numpy(weights_output_layer)
+
+        if self.use_bias:
+            output_bias, index = self.read_matrix_from_genome(individual, index, self.output_size, 1)
+            self.output_layer.bias.data = torch.from_numpy(output_bias.flatten())
 
     def brainstep(self, ob: np.ndarray) -> np.ndarray:
+
+        assert ob.ndim == 1
+        assert ob.size == self.input_size
```

Comment:
This is actually not a problem since the changes from #53 guarantee that only a NumPy array will be passed to `brainstep()`. So the brains can be sure that they get a NumPy array but what exactly is inside the NumPy Array needs to be handled by the brain.

So this will only create trouble when for example an RGB input is parsed but that's okay if that fails here because I think the brain should decide if it accepts multi-dimensional input like RGB or not.